ocaml/testsuite/tests/typing-layouts/rep_poly_native.ml

```diff
@@ -0,0 +1,60 @@
+(* TEST
+   flags = "-extension layouts"
+   * flambda2
+   ** native
+   ** bytecode
+*)
+
+
+external[@rep_poly] id : ('a : any). 'a -> 'a = "%identity"
+external[@rep_poly] opaque : ('a : any). 'a -> 'a = "%opaque"
+external[@rep_poly] magic : ('a : any) ('b : any). 'a -> 'b = "%obj_magic"
```

Comment:
Shouldn't there be a constraint for `magic` that both `'a` and `'b` have the same layout?

Reply:
https://github.com/ocaml-flambda/flambda-backend/blob/213a02303b0501345b3559fbc054879e1b74d644/ocaml/testsuite/tests/typing-layouts/rep_poly.ml#L363-L380

Yes, for the current implementation, all the type variables with layout `any` in the `[@rep_poly]` external declaration will have to have the same representable layout per call site.

Seperate from this feature, something like `external magic : float# -> int32# = "%obj_magic"` would still be accepted at the moment. Wonder if we should find a way to ban it.

Reply:
There is no typechecking for primitives at all today.  So you can write:
```ocaml
external eq : 'a -> 'b -> string = "%equal"
 
let _ = Printf.printf "string? %s" (eq 42 true)
```
and it just segfaults.  So the situation for layouts is sad, but not a new problem with layouts.  This is tracked as 1335 in our internal ticket system, but there is no plan to do anything about it soon.

Reply:
I think we still need actual checking, as the issue is not of whether the produced code segfaults, but of whether the compiler crashes. It's independent, though, since things like `external magic : float# -> int32# = "%obj_magic"` is indeed accepted at the moment, and is likely to crash flambda2 (or some slightly different version of it if it doesn't; I'm almost sure that `external magic : (int, int) unboxed_pair -> int = "%obj_magic"` will causes crashes because we won't know what to do when doing unarization).

Reply:
Added a commit to strengthen checks around this.

https://github.com/ocaml-flambda/flambda-backend/blob/83c13fa212e12709fb70bd213521cd24d84421b0/ocaml/typing/primitive.ml#L451-L546

It restricts the set of built-in primitives that can have argument/return types of non-value jkinds. And for those that are allowed, it does some sanity checks. It able to make sure the argument and return representation of "%obj_magic" are the same, "%box_float" is not declared to take `int32#`, etc.